File: app/src/main/java/com/brainor/bilihelper/HistorySettingsActivity.java
package com.brainor.bilihelper;

import android.content.Intent;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.Objects;

public class HistorySettingsActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.history_settings);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);

        RecyclerView HistoryListView = findViewById(R.id.HistoryList);
        HistoryListView.setLayoutManager(new LinearLayoutManager(this));
        MyAdapter adapter = new MyAdapter();
        HistoryListView.setAdapter(adapter);
        (new ItemTouchHelper(new SwipeToDeleteCallback(adapter))).attachToRecyclerView(HistoryListView);
    }

    class MyAdapter extends RecyclerView.Adapter<MyAdapter.MyViewHolder> {
        @NonNull
        @Override
        public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            TextView v = (TextView) LayoutInflater.from(parent.getContext()).inflate(R.layout.support_simple_spinner_dropdown_item, parent, false);
            TypedValue outValue = new TypedValue();
            getApplicationContext().getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue, true);
            v.setBackgroundResource(outValue.resourceId);
            return new MyViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
            holder.mTextView.setText(MainActivity.HistoryList.get(position).title);
            holder.mTextView.setOnClickListener(v ->
                    getApplication().startActivity(new Intent(getApplicationContext(), MainActivity.class).putExtra("position", holder.getAdapterPosition()))
            );
        }

        @Override
        public int getItemCount() {
            return MainActivity.HistoryList.size();
        }


        class MyViewHolder extends RecyclerView.ViewHolder {
            TextView mTextView;

            MyViewHolder(TextView v) {
                super(v);
                mTextView = v;
            }

        }

        void deleteItem(int position) {
            MainActivity.HistoryList.remove(position);
            notifyItemRemoved(position);
            MainActivity.StoreHistory(getApplicationContext());
        }
    }

    class SwipeToDeleteCallback extends ItemTouchHelper.SimpleCallback {
        private MyAdapter mAdapter;
        private Drawable icon;
        private final ColorDrawable background;
        SwipeToDeleteCallback(MyAdapter adapter) {
            super(0, ItemTouchHelper.LEFT);
            mAdapter = adapter;
            icon = ContextCompat.getDrawable(getApplicationContext(), android.R.drawable.ic_menu_delete);
            background = new ColorDrawable(Color.RED);
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
            int position = viewHolder.getAdapterPosition();
            mAdapter.deleteItem(position);
        }

        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder viewHolder1) {
            return false;
        }

        @Override
        public void onChildDraw(@NonNull Canvas c, @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
            View itemView = viewHolder.itemView;
            int backgroundCornerOffset = 0;
            if (dX < 0) {
                background.setBounds(itemView.getRight() + ((int) dX) - backgroundCornerOffset, itemView.getTop(), itemView.getRight(), itemView.getBottom());
            } else background.setBounds(0, 0, 0, 0);
            background.draw(c);
            int iconMargin = (itemView.getHeight() - icon.getIntrinsicHeight()) / 2;
            int iconTop = itemView.getTop() + (itemView.getHeight() - icon.getIntrinsicHeight()) / 2;
            int iconBottom = iconTop + icon.getIntrinsicHeight();
            if (dX < 0) { // Swiping to the left
                int iconLeft = itemView.getRight() - iconMargin - icon.getIntrinsicWidth();
                int iconRight = itemView.getRight() - iconMargin;
                icon.setBounds(iconLeft, iconTop, iconRight, iconBottom);

                background.setBounds(itemView.getRight() + ((int) dX) - backgroundCornerOffset,
                        itemView.getTop(), itemView.getRight(), itemView.getBottom());
            } else { // view is unSwiped
                background.setBounds(0, 0, 0, 0);
            }
            background.draw(c);
            icon.draw(c);
        }

    }
}

File: app/src/main/java/com/brainor/bilihelper/HistorySettingsActivity.kt
package com.brainor.bilihelper

import android.content.Intent
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.history_settings.*

class HistorySettingsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.history_settings)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        historyListView.layoutManager = LinearLayoutManager(this)
        val adapter = MyAdapter()
        historyListView.adapter = adapter
        ItemTouchHelper(SwipeToDeleteCallback(adapter)).attachToRecyclerView(historyListView)
    }

    internal inner class MyAdapter : RecyclerView.Adapter<MyAdapter.MyViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.support_simple_spinner_dropdown_item, parent, false) as TextView
            val outValue = TypedValue()
            applicationContext.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
            v.setBackgroundResource(outValue.resourceId)
            return MyViewHolder(v)
        }

        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            holder.mTextView.text = MainActivity.HistoryList[position].title
            holder.mTextView.setOnClickListener {  application.startActivity(Intent(applicationContext, MainActivity::class.java).putExtra("position", holder.adapterPosition)) }
        }

        override fun getItemCount(): Int {
            return MainActivity.HistoryList.size
        }


        internal inner class MyViewHolder(var mTextView: TextView) : RecyclerView.ViewHolder(mTextView)

        fun deleteItem(position: Int) {
            MainActivity.HistoryList.removeAt(position)
            notifyItemRemoved(position)
            MainActivity.storeHistory(applicationContext)
        }
    }

    internal inner class SwipeToDeleteCallback(private val mAdapter: MyAdapter) : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
        private val icon: Drawable = ContextCompat.getDrawable(applicationContext, android.R.drawable.ic_menu_delete)!!
        private val background: ColorDrawable = ColorDrawable(Color.RED)

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, i: Int) {
            val position = viewHolder.adapterPosition
            mAdapter.deleteItem(position)
        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, viewHolder1: RecyclerView.ViewHolder): Boolean {
            return false
        }

        override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
            val itemView = viewHolder.itemView
            val backgroundCornerOffset = 0
            if (dX < 0) {
                background.setBounds(itemView.right + dX.toInt() - backgroundCornerOffset, itemView.top, itemView.right, itemView.bottom)
            } else
                background.setBounds(0, 0, 0, 0)
            background.draw(c)
            val iconMargin = (itemView.height - icon.intrinsicHeight) / 2
            val iconTop = itemView.top + (itemView.height - icon.intrinsicHeight) / 2
            val iconBottom = iconTop + icon.intrinsicHeight
            if (dX < 0) { // Swiping to the left
                val iconLeft = itemView.right - iconMargin - icon.intrinsicWidth
                val iconRight = itemView.right - iconMargin
                icon.setBounds(iconLeft, iconTop, iconRight, iconBottom)

                background.setBounds(itemView.right + dX.toInt() - backgroundCornerOffset,
                        itemView.top, itemView.right, itemView.bottom)
            } else { // view is unSwiped
                background.setBounds(0, 0, 0, 0)
            }
            background.draw(c)
            icon.draw(c)
        }

    }
}